test/distributed/test_c10d_nccl.py

```diff
@@ -2454,6 +2454,199 @@ def test_channels_last_contig(self):
         process_group.broadcast([tensor]).wait()
 
 
+# These are internal OpType enum values in Work.hpp, the value is subject to
+# change in future versions.
+BROADCAST = 0
+ALLREDUCE = 1
+ALLGATHER = 4
+
+
+class WorkHookTest(MultiProcessTestCase):
+
+    @property
+    def world_size(self):
+        return 2
+
+    def setUp(self):
+        super().setUp()
+        # set NCCL_ENABLE_TIMING to enable timing for CUDAEvents
+        # in ProcessGroup Work
+        os.environ["NCCL_ENABLE_TIMING"] = "1"
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def _get_store(self):
+        return dist.FileStore(self.file_name, self.world_size)
+
+    def _get_process_group(self):
+        store = self._get_store()
+        c10d.init_process_group("nccl", store=store, rank=self.rank, world_size=self.world_size)
+        return c10d.distributed_c10d._get_default_group()
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_on_completion_hook_broadcast(self):
+        pg = self._get_process_group()
+        num_hook_fired = 0
+        durations: List[float] = []
+
+        def hook(work):
+            nonlocal num_hook_fired, durations
+            num_hook_fired += 1
+            durations.append(work._get_duration())
+
+        pg._register_on_completion_hook(hook)
+        tensor = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        pg.broadcast([tensor]).wait()
+        pg.broadcast([tensor]).wait()
+
+        # N.B.: destroy_process_group is necessary to wait for
+        # all pending works to finish.
+        c10d.destroy_process_group(pg)
+
+        self.assertEqual(num_hook_fired, 2)
+        self.assertEqual(len(durations), 2)
+        for duration in durations:
+            self.assertTrue(duration > 0)
+
+        self.assertEqual(tensor, torch.zeros([2, 3]).cuda(self.rank))
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_on_completion_hook_mixed_ops(self):
+        pg = self._get_process_group()
+        num_hook_fired = 0
+        durations: List[float] = []
+
+        def hook(work):
+            nonlocal num_hook_fired, durations
+            num_hook_fired += 1
+            durations.append(work._get_duration())
+
+        pg._register_on_completion_hook(hook)
```

Comment:
building on your PR desc saying that trainer may not have insight into model / collectives used, should we assume the same thing about PG creation?  I wonder if its valuable to have a higher-level hook installed globally on c10d lib by the trainer that would log work events across PGs.

(if you agree, i think this could still be a separate PR and probably build on top of the infra in this PR)

Reply:
> building on your PR desc saying that trainer may not have insight into model / collectives used, should we assume the same thing about PG creation? 

Good point! It's less intrusive to get PGs compared to every Work, but it's certainly better if infra/trainers do not have to. 

> I wonder if its valuable to have a higher-level hook installed globally on c10d lib by the trainer that would log work events across PGs.

Sure, I can add that in a followup PR.